analyzers/src/SonarAnalyzer.Common/Rules/ExpressionComplexityBase.cs

```diff
@@ -30,39 +30,57 @@ public abstract class ExpressionComplexityBase<TExpression> : ParameterLoadingDi
+        protected abstract TSyntaxKind[] TransparentKinds { get; }
+        protected abstract SyntaxNode[] ExpressionChildren(SyntaxNode node);
 
         [RuleParameter("max", PropertyType.Integer, "Maximum number of allowed conditional operators in an expression", DefaultValueMaximum)]
         public int Maximum { get; set; } = DefaultValueMaximum;
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
 
         protected ExpressionComplexityBase() =>
-            rule = Language.CreateDescriptor(DiagnosticId, MessageFormat, isEnabledByDefault: false);
+            rule ??= Language.CreateDescriptor(DiagnosticId, MessageFormat, isEnabledByDefault: false);
 
         protected sealed override void Initialize(ParameterLoadingAnalysisContext context) =>
-            context.RegisterSyntaxTreeActionInNonGenerated(Language.GeneratedCodeRecognizer, c =>
+            context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer, c =>
                 {
-                    var root = c.Tree.GetRoot();
-                    var rootExpressions = NoncompoundSubexpressions(root);
-                    var compoundExpressionsDescendants = root.DescendantNodes().Where(IsCompoundExpression).SelectMany(NoncompoundSubexpressions);
-
-                    foreach (var expression in rootExpressions.Concat(compoundExpressionsDescendants))
+                    if (NodeKind(c.Node.Parent) is TSyntaxKind parentKind && (ComplexityIncreasingKinds.Contains(parentKind) || TransparentKinds.Contains(parentKind)))
+                    {
+                        // The parent of the expression is itself complexity increasing (e.g. &&) or transparent (e.g. parenthesis).
+                        // We are only interested in the expression roots so we only report once per expression tree. Therefore we ignore any inner children, e.g.:
+                        // a && b && c -> (Left: (Left: a, Right: b), Right: c) // We are only interested in the outer expression (the one with Right: c)
+                        return;
+                    }
+                    var complexity = CalculateComplexity(c.Node);
+                    if (complexity > Maximum)
                     {
-                        var complexity = expression.DescendantNodesAndSelf(x => !IsCompoundExpression(x)).Count(IsComplexityIncreasingKind);
-                        if (complexity > Maximum)
-                        {
-                            c.ReportIssue(Diagnostic.Create(rule, expression.GetLocation(), Maximum, complexity));
-                        }
+                        c.ReportIssue(Diagnostic.Create(rule, c.Node.GetLocation(), Maximum, complexity));
                     }
-                });
+                }, ComplexityIncreasingKinds.Concat(TransparentKinds).ToArray());
+
+        private static TSyntaxKind? NodeKind(SyntaxNode node) =>
+            node == null
```

Comment:
Educational: is this a defensive-coding check? If it's not maybe we can add a test case to increase the coverage?

Reply:
It kind of is but it is untestable. We have one `NodeKind(c.Node.Parent)` call which in theory could be null but the expressions kinds we register for at the moment have a parent always.